### PR TITLE
Update product index if category products updated

### PR DESCRIPTION
**Pull request by thaiphan**
This change enables the Algolia extension for Magento to now automatically update the product index when a product is added to or removed from a category via the "Manage Categories" menu.

This change also moves some duplicate code into a helper class.

### CODE REVIEW

File: code/Helper/Indexer.php [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+     * This function will update all the requested categories and their child
+     * categories in Algolia. You can provide either a single category ID or an
+     * array of category IDs. A category ID should be either a string or an
+     * integer.
+     *
+     * @param array|string|int $updateCategoryIds
+     */
+    public function reindexSpecificCategories($updateCategoryIds)
+    {
+        $updateCategoryIds = is_array($updateCategoryIds) ? $updateCategoryIds : [$updateCategoryIds];
```
Comment (JanPetr) [inline code/Helper/Indexer.php]: This can be shorten:
```php 
$updateCategoryIds = (array) $updateCategoryIds;
```

File: code/Helper/Indexer.php [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    {
+        return Mage::getResourceSingleton('catalogsearch/indexer_fulltext');
+    }
+
+    /**
+     * @return Algolia_Algoliasearch_Model_Resource_Engine
+     */
+    protected function getEngine()
+    {
+        if (is_null($this->engine)) {
```
Comment (JanPetr) [inline code/Helper/Indexer.php]: I would prefer not to set `NULL` to `$engine` variable and testing `!isset($this->engine)`, but that's only cosmetic update

File: code/Model/Indexer/Algolia.php [hunk 3]
```diff
@@ -13,6 +13,9 @@ class Algolia_Algoliasearch_Model_Indexer_Algolia extends Mage_Index_Model_Index
     /** @var Algolia_Algoliasearch_Helper_Logger */
     protected $logger;
 
+    /** @var Algolia_Algoliasearch_Helper_Indexer */
+    protected $indexer;
```
Comment (JanPetr) [inline code/Model/Indexer/Algolia.php]: Woul be better to expand the var name to something like `$indexHelper`. Only `$indexer` might be confusing with `Indexer` classes.

File: code/Model/Indexer/Algoliacategories.php [hunk 4]
```diff
@@ -10,6 +10,9 @@ class Algolia_Algoliasearch_Model_Indexer_Algoliacategories extends Mage_Index_M
     /** @var Algolia_Algoliasearch_Helper_Config */
     protected $config;
 
+    /** @var Algolia_Algoliasearch_Helper_Indexer */
+    protected $indexer;
```
Comment (JanPetr) [inline code/Model/Indexer/Algoliacategories.php]: The same with the var name as in `Algolia.php`.

File: code/Model/Indexer/Algoliacategories.php [hunk 5]
```diff
@@ -156,18 +152,15 @@ protected function _processEvent(Mage_Index_Model_Event $event)
...
-                    $updateCategoryIds[] = $category->getId();
-                }
-            }
+            $this->indexer->reindexSpecificCategories($data['catalogsearch_update_category_id']);
+        }
 
-            $this->engine->rebuildCategoryIndex(null, $updateCategoryIds);
+        /**
+         * If we have added any new products to a category then we need t
+         * update that product in Algolia to reflect it.
```
Comment (JanPetr) [inline code/Model/Indexer/Algoliacategories.php]: `If we have added any new products to a category then we need to update these products in Algolia indices`
Wouldn't it be better?

File: code/Model/Indexer/Algolia.php [hunk 6]
```diff
@@ -171,14 +175,6 @@ protected function _registerCatalogProductEvent(Mage_Index_Model_Event $event)
         return $this;
     }
 
-    protected function _isProductComposite($productId)
```
Comment (JanPetr) [inline code/Model/Indexer/Algolia.php]: I'm not sure we want to remove the method completely. It definitelly makes sense to move the method to one class but I'm worried about such BC break. I am thinking about keeping the method and just call `$this->index->isProductComposite($productId);` from it and trigger `Deprecated` warning? WDYT?

**Comment by JanPetr** [comment]
@thaiphan Thank you for the PR, from my POV makes sense. I just added few comments to some lines. And I will invite @maxiloc to take a look.

**Comment by JanPetr** [comment]
OK, we discussed it and the "helper" approach is not the ideal one. It would be better if you create a new abstract indexer class and move the duplicate code there. It would keep the flow of data and do not cause BC breaks (like with `_isProductComposite` method).

Are you able to refactor it?

**Comment by thaiphan** [comment]
@JanPetr Okay.

**Comment by thaiphan** [comment]
@JanPetr @maxiloc Okay, I updated to use an abstract class. Can you review and merge?

**Comment by JanPetr** [comment]
@thaiphan Look way better, thanks! Can you just squash your commits and fix StyleCI issues?

**Comment by thaiphan** [comment]
@JanPetr I squashed everything but I'm confused about StyleCI. I'm pretty sure I already made the suggested change.

**Pull request merged**